src/Hamiltonians/FroelichPolaron_nD.jl

```diff
@@ -0,0 +1,235 @@
+        address,D, geometry, alpha, mass, omega, l, p, ks, momentum_cutoff, mode_cutoff)
+end
+
+
+function Base.show(io::IO, h::FroehlichPolaronND)
+    io = IOContext(io, :compact => true)
+    println(io, "FroehlichPolaronND(")
+    println(io, "  ", starting_address(h), ",")
+    println(io, "  geometry = ", h.geometry, ",")
+    println(io, "  alpha = ", h.alpha, ", mass = ", h.mass, ", omega = ", h.omega, ", Dimention = ", h.d,",")
+    println(io, "  l = ", Float64.(h.l), ", p = ", Float64.(h.p), ",")
+    !isnothing(h.momentum_cutoff) && println(io, "  momentum_cutoff = ", h.momentum_cutoff, ",")
+    println(io, "  mode_cutoff = ", isnothing(h.mode_cutoff) ? "nothing" : string(h.mode_cutoff))
+    print(io, ")")
+end
+
+LOStructure(::Type{<:FroehlichPolaronND}) = IsHermitian()
+
+starting_address(h::FroehlichPolaronND) = h.address
+
+function dimension(h::FroehlichPolaronND, address)
+    # takes into account `mode_cutoff` but not `momentum_cutoff`
+    M = num_modes(address)
+    n = h.mode_cutoff
+    return BigInt(n + 1)^BigInt(M)
+end
+
+struct FroehlichPolaronNDColumn{H,A} <: AbstractOperatorColumn{A,Float64,H}
+    hamiltonian::H
+    address::A
+    num_offdiagonals::Int
+end
+
+
+
+function diagonal_element(col::FroehlichPolaronNDColumn)
+    h = col.hamiltonian
+    occ = onr(col.address)
+
+    Pphonon = zero(h.ks[1])
+    Nphonon = 0
+    for m in 1:num_modes(col.address)
+        nm = occ[m]
+        Nphonon += nm
+        Pphonon += h.ks[m] * nm
+    end
+    ek = dot(h.p - Pphonon, h.p - Pphonon) / (h.mass)
+    return h.omega * Nphonon + ek
+end
+
+
+struct FroehlichPolaronNDOffdiagonals{A,H} <: AbstractVector{Pair{A,Float64}}
+    address::A
+    h::H
+    num_offdiagonals::Int
+end
+
+offdiagonals(col::FroehlichPolaronNDColumn) = FroehlichPolaronNDOffdiagonals(col.address, col.hamiltonian, col.num_offdiagonals)
+
+
+Base.size(ods::FroehlichPolaronNDOffdiagonals) = (ods.num_offdiagonals,)
+
+
+@inline function calc_vk(h::FroehlichPolaronND, kidx::Int)
+    knorm = sqrt(dot(h.ks[kidx], h.ks[kidx]))
+    vol = prod(h.l)
+    if h.d == 1
+        return (2 * h.alpha /(h.l[1]))^0.5
+    else
+        if knorm == 0.0
+            return 0.0
+        else
+            return (-1) * sqrt((gamma(((h.d-1)/2)) * h.alpha * 2^(h.d-(3/2)) * pi^((h.d-1)/2) * h.omega^2) / (sqrt(h.mass * h.omega) * (knorm^(h.d-1) * vol)))
+        end
+    end
+end
+
+@inline function phonon_op(h::FroehlichPolaronND, addr, chosen)
+    M = num_modes(addr)
+    if chosen ≤ M
+        if !isnothing(h.mode_cutoff) && onr(addr)[chosen] ≥ h.mode_cutoff
+            return addr => 0.0
+        end
+        new_addr, val = excitation(addr, (chosen,), ())
+        vk = calc_vk(h, chosen)
+        amp = -vk * val
+        
+    else
+        k = chosen - M
+        new_addr, val = excitation(addr, (), (k,))
+        vk = calc_vk(h, k)
+        amp = -vk * val
+        
+    end
+    
+
+    if h.momentum_cutoff != nothing
```

Comment:
```suggestion
    if !isnothing(h.momentum_cutoff)
```
This is slightly more efficient.